Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionControllerCocoa.mm

```diff
@@ -58,10 +58,59 @@
 #import <wtf/NeverDestroyed.h>
 #import <wtf/text/WTFString.h>
 
+#if PLATFORM(IOS_FAMILY)
+#import "FoundationSPI.h"
+#endif
+
 static constexpr Seconds purgeMatchedRulesInterval = 5_min;
 
+static NSString * const WebExtensionsUniqueIdentifierKey = @"uniqueIdentifier";
+static NSString * const WebExtensionsLocalStorageWasDeletedNotification = @"WebExtensionsLocalStorageWasDeleted";
+
+using namespace WebKit;
+
+@interface _WKWebExtensionControllerHelper : NSObject {
+    WeakPtr<WebKit::WebExtensionController> _webExtensionController;
+}
+
+- (instancetype)initWithWebExtensionController:(WebKit::WebExtensionController&)controller;
+
+@end
+
+@implementation _WKWebExtensionControllerHelper
+
+- (instancetype)initWithWebExtensionController:(WebKit::WebExtensionController&)controller
+{
+    if (!(self = [super init]))
+        return nil;
+
+    _webExtensionController = controller;
+
+    [NSDistributedNotificationCenter.defaultCenter addObserver:self selector:@selector(_didDeleteLocalStorage:) name:WebExtensionsLocalStorageWasDeletedNotification object:nil];
+
+    return self;
+}
+
+- (void)_didDeleteLocalStorage:(NSNotification *)notification
+{
+    NSString *uniqueIdentifier = notification.userInfo[WebExtensionsUniqueIdentifierKey];
+    if (!uniqueIdentifier)
+        return;
+
+    if (RefPtr context = _webExtensionController->extensionContext(uniqueIdentifier))
+        context->invalidateStorage();
+}
+
+@end
+
 namespace WebKit {
 
+void WebExtensionController::createWebExtensionControllerHelper()
+{
+    ASSERT(!m_webExtensionControllerHelper);
+    m_webExtensionControllerHelper = [[_WKWebExtensionControllerHelper alloc] initWithWebExtensionController:*this];
```

Comment:
`adoptNS([[_WKWebExtensionControllerHelper alloc] initWithWebExtensionController:*this]);`

Reply:
That isn't needed here, we use ARC for these extension files.